---
title: "gene expression"
output:
  word_document: default
  html_document: default
---

```{r init}
library(Seurat)
library(ggplot2)
library(patchwork)
library(data.table)

# Load the dataset and remove cells with identity NA
calr <- readRDS("/Users/lizacasella/Desktop/Nam_Lab/calr_tcr.rds")


# subset to remove cells with seurat cluster or IFN status = NA
calr <- subset(calr, subset = seurat_clusters != "NA")
calr <- subset(calr, subset = IFN.Status2 != "NA")

```


```{r}

# preprocess and run PCA
calr <- NormalizeData(calr)
calr <- FindVariableFeatures(calr)
calr <- ScaleData(calr)
calr <- RunPCA(calr, verbose = FALSE)

# plot by seurat clusters 
DimPlot(calr,group.by = "seurat_clusters", label = TRUE, label.size = 6)

# plot by cell identity 
DimPlot(calr,group.by = "Cluster.Ident")

# plot seurat clusters seperatly, pre- and post-IFN
DimPlot(calr,group.by = "seurat_clusters", split.by = "IFN.Status2", label = TRUE, label.size = 4)
 
```


```{r}

# find genes differentially expressed between clusters with a >2 fold increase 
# between average expression 

# find number/names of clusters 
cluster <- levels(calr)
num = length(levels(calr))

# choose how many differentially expressed genes you want to see
num_genes <- 8

# make a data frame for the genes from each cluster
# (it doesn't work)
genes <- data.frame(matrix(nrow = num_genes,ncol=num))
colnames(genes) <- cluster


# find and print differentially expressed genes 
for (i in 0:15) {
  seurat <- cluster[i+1]
  
  # find genes
  markers <- FindMarkers(calr, ident.1 = seurat, ident.2 = NULL, logfc.threshold = log(2))
  
  # change them from row names to a column
  markers <- setDT(markers, keep.rownames = seurat)
  
  # isolate just the gene names
  markers <- data.frame(markers[1:num_genes, ])
  
  # print the cluster number and DE genes 
  print(paste0("cluster ", i))
  print(markers)
}


```




