---
title: "pseudotime"
output: html_document
---

```{r init}
library(dplyr) # data manipulation
library(Seurat) # toolkit single cell RNA seq data
library(monocle3)
```

```{r}

calr <- readRDS("/Users/lizacasella/Desktop/Nam_Lab/calr.3p.rds")

gene_annotation <- as.data.frame(rownames(calr@reductions[["pca"]]@feature.loadings), row.names = rownames(calr@reductions[["pca"]]@feature.loadings))
colnames(gene_annotation) <- "gene_short_name"

# part two, cell information

cell_metadata <- as.data.frame(calr@assays[["RNA"]]@counts@Dimnames[[2]], row.names = calr@assays[["RNA"]]@counts@Dimnames[[2]])
colnames(cell_metadata) <- "barcode"

# part three, counts sparse matrix

New_matrix <- calr@assays[["RNA"]]@counts
New_matrix <- New_matrix[rownames(calr@reductions[["pca"]]@feature.loadings), ]
expression_matrix <- New_matrix

cds <- new_cell_data_set(expression_matrix,
                                     cell_metadata = cell_metadata,
                                     gene_metadata = gene_annotation)

cell <- data.frame(calr@meta.data$Cluster.Ident_IFN_Geno)
ident <- data.frame(calr@meta.data$Cluster.Ident)
seurat <- data.frame(calr@meta.data$seurat_clusters)
ifn <- data.frame(calr@meta.data$IFN.Status2)
mut <- data.frame(calr@meta.data$Genotype)
t <- data.frame(calr@meta.data$hash.ID)


cds@colData$cell <- calr@meta.data$Cluster.Ident_IFN_Geno
cds@colData$ifn <- calr@meta.data$IFN.Status
cds@colData$geno <- calr@meta.data$Genotype
cds@colData$ident <- calr@meta.data$Cluster.Ident
cds@colData$seurat <- calr@meta.data$seurat_clusters

```

```{r}
cds <- preprocess_cds(cds, num_dim = 50)
cds <- reduce_dimension(cds)
#plot_cells(cds, label_groups_by_cluster=FALSE,  color_cells_by = "ident")

cds <- cluster_cells(cds)
#plot_cells(cds, color_cells_by = "partition")

cds <- learn_graph(cds)
#plot_cells(cds, color_cells_by = "ident", label_groups_by_cluster=FALSE, label_leaves=FALSE, label_branch_points=FALSE)

#plot_cells(cds, color_cells_by = "ifn", label_cell_groups=FALSE, label_leaves=TRUE, label_branch_points=TRUE, graph_label_size=1.5)

cds <- order_cells(cds)

plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5)
```

```{r}

# assigns pseudotime values to every cell 
traj.coord <- cds@principal_graph_aux@listData[["UMAP"]][["pseudotime"]]
time <- data.frame(traj.coord)
time$cell <- calr@meta.data$Cluster.Main
time$gene <- calr@meta.data$Genotype

```




